agent/php_environment.c

```diff
@@ -511,6 +511,94 @@ static void nr_php_get_environment_variables(TSRMLS_D) {
                    __func__, NR_PHP_PROCESS_GLOBALS(env_labels));
 }
 
+#define DOCKER_ID_V2_STRLEN (64)
+/*
+ * Purpose:
+ *    Extract the 64-byte hexadecimal Docker cgroup ID from
+ *     /proc/self/mountinfo
+ */
+void nr_php_get_v2_docker_id(const char* cgroup_fname) {
+  char* line_ptr = NULL;
+  char* token = NULL;
+  bool found = false;
+  FILE* fd = NULL;
+  size_t len = 0;
+  nr_regex_t* regex = NULL;
+
+  // check if docker_id global already set
+  if (NULL != NR_PHP_PROCESS_GLOBALS(docker_id)) {
+    nrl_verbosedebug(NRL_AGENT, "Docker ID already set.");
+    return;
+  }
+
+  // check if file exists
+  if (SUCCESS != access(cgroup_fname, F_OK)) {
+    nrl_verbosedebug(NRL_AGENT, "File not found: %s", cgroup_fname);
+    return;
+  }
+
+  // open file
+  fd = fopen(cgroup_fname, "r");
+  if (NULL == fd) {
+    nrl_warning(NRL_AGENT, "Failed to open %s", cgroup_fname);
+    return;
+  }
+
+  // compile regex to verify hexadecimal chars
+  regex = nr_regex_create("^[a-fA-F0-9]+$", 0, 0);
+
+  // clang-format off
+  /* 
+   * Example /proc/self/mountinfo file structure:
+   *   ...
+   *   795 787 254:1 /docker/containers/ec807d5258c06c355c07e2acb700f9029d820afe5836d6a7e19764773dc790f5/resolv.conf /etc/resolv.conf rw,relatime - ext4 /dev/vda1 rw
+   *   796 787 254:1 /docker/containers/ec807d5258c06c355c07e2acb700f9029d820afe5836d6a7e19764773dc790f5/hostname /etc/hostname rw,relatime - ext4 /dev/vda1 rw
+   *   797 787 254:1 /docker/containers/ec807d5258c06c355c07e2acb700f9029d820afe5836d6a7e19764773dc790f5/hosts /etc/hosts rw,relatime - ext4 /dev/vda1 rw
+   *   ...
+   */
+
+  /*
+   * File parsing logic:
+   *  1. scan file line-by-line
+   *  2. split each line into '/' delimited tokens
+   *  3. search for "docker" followed by "container" string tokens
+   *  4. extract the 64 byte field following "/docker/containers/"
+   *  5. Verify the extracted ID is both:
+   *    a. 64 bytes long (not including null terminator)
+   *    b. comprised of only hexadecimal characters
+   *  6. Assign the extracted & verified ID to the global
+   *    a. Example ID: ec807d5258c06c355c07e2acb700f9029d820afe5836d6a7e19764773dc790f5
+   *  7. Set found = true and exit the loops
+   */
+  // clang-format on
+
+  nrl_verbosedebug(NRL_AGENT, "Reading %s:", cgroup_fname);
+
+  while (FAILURE != getline(&line_ptr, &len, fd) && !found) {
```

Comment:
So this code will loop infinitely if for some reason the source continues to provide data - say "/dev/zero".  Understandably the file we're looking at shouldn't do that.  But would it be worth putting in some kind of limiter so if something goes wrong the entire PHP process doesn't lock up?  Like a maximum number of passes through the loop?

Reply:
I can add code to set an arbitrary loop count to bail out after X number of loops. How about 100? I don't think most mountinfo files will have more lines than that.

Reply:
Added in d7039ee8dc39600064ce9631f28aa86d0d3ef600

Reply:
Maybe higher - on my linux machine at home has 73 lines in it.  With more software coming in 'snap' format I believe this can cause a large number of mounts.  I think 1000 would be fine - it should never be reached unless something is probably very wrong with `/proc/self/mountinfo`.   Adding a log message if it bails out for this reason would be helpful in diagnosing the (rare hopefully) situation this limit is reached.

Reply:
Upped this value in b878efa816b73ccf9c186a00b4b9086450582bb8
Also took the opportunity to add a boundary to the inner loop with a max of 10 token parsing loops. The actual field we want shouldn't take more than 2 on a standard system, but just in case this file format changes in the future or a user has an unusual setup we'll have a reasonable buffer.

Reply:
Good idea!